sky/skylet/providers/kubernetes/node_provider.py

```diff
@@ -237,10 +239,42 @@ def create_node(self, node_config, tags, count):
                         # If container_statuses is None, then the pod hasn't
                         # been scheduled yet.
                         all_ready = False
-            if all_ready:
+
+            # check if all the pods and containers within the pods are running
+            if  all([ pod.status.phase == "Running" for pod in pods]) \
+                and all([container.state.running for pod in pods for container in pod.status.container_statuses]):
+                pods_and_containers_running = True
+
+            if all_ready and pods_and_containers_running:
                 break
             time.sleep(1)
 
+        # Kubernetes automatically populates containers with critical
+        # environment variables, such as those for discovering services running
+        # in the cluster and CUDA/nvidia environment variables. We need to
+        # update task environment variables with these env vars. This is needed
+        # for GPU support and service discovery.
+        # See https://github.com/skypilot-org/skypilot/issues/2287 for
+        # more details.
+        # Capturing env. var. from the pod's runtime and writes them to
+        # /etc/profile.d/ making them available for all users in future
+        # shell sessions.
+        set_k8s_env_var_cmd = [
+            '/bin/sh', '-c',
+            'printenv | awk \'{print "export " $0}\' > ~/k8s_env_var.sh && sudo mv ~/k8s_env_var.sh /etc/profile.d/k8s_env_var.sh'
```

Comment:
Bug: After the first run, `bash: 419: No such file or directory` gets printed every time I SSH or exec a task.

Repro:
1. `sky launch -c test --gpus T4:1 -- nvidia-smi`
2. `sky exec test --gpus T4:1 -- nvidia-smi` (This prints the above error). [Logs](https://gist.github.com/romilbhardwaj/76c5e0cbfc59623d4e9f3b1f5776ade7).
3. `ssh test` this also prints the above error.

Cause:
`/etc/profile.d/k8s_env_var.sh` contains a line like:
```
export NVIDIA_REQUIRE_CUDA=cuda>=11.6 brand=tesla,driver>=418,driver<419 brand=tesla,driver>=450,driver<451 brand=tesla,driver>=470,driver<471 brand=unknown,driver>=470,driver<471 brand=nvidia,driver>=470,driver<471 brand=nvidiartx,driver>=470,driver<471 brand=geforce,driver>=470,driver<471 brand=geforcertx,driver>=470,driver<471 brand=quadro,driver>=470,driver<471 brand=quadrortx,driver>=470,driver<471 brand=titan,driver>=470,driver<471 brand=titanrtx,driver>=470,driver<471
```
The shell tries to interpret special characters > and <, causing errors.

Solution - wrap the exported value in single quotes:
```
export NVIDIA_REQUIRE_CUDA='cuda>=11.6 brand=tesla,driver>=418,driver<419 brand=tesla,driver>=450,driver<451 brand=tesla,driver>=470,driver<471 brand=unknown,driver>=470,driver<471 brand=nvidia,driver>=470,driver<471 brand=nvidiartx,driver>=470,driver<471 brand=geforce,driver>=470,driver<471 brand=geforcertx,driver>=470,driver<471 brand=quadro,driver>=470,driver<471 brand=quadrortx,driver>=470,driver<471 brand=titan,driver>=470,driver<471 brand=titanrtx,driver>=470,driver<471'
```

Catch - if the envvar var value itself contains single quotes, we'll need to be careful in handling single quotes (`'`) appearing in the string. See some suggestions [here](https://stackoverflow.com/a/1315213).